### PR TITLE
Do not timeout watch requests

### PR DESCRIPTION
**Pull request by nevalla**
Watch requests are closed by `socket hangup` because  api proxy has 30 seconds timeout set. We need to disable the timeout by setting it 0 for watch requests. However, `http-proxy` package has a bug that prevents it to set as 0, so I had to patch the implementation.

### CODE REVIEW

File: patches/http-proxy+1.17.0.patch [hunk 1]
```diff
@@ -0,0 +1,22 @@
+diff --git a/node_modules/http-proxy/lib/http-proxy/passes/web-incoming.js b/node_modules/http-proxy/lib/http-proxy/passes/web-incoming.js
+index 995a0db..a5c37f5 100644
+--- a/node_modules/http-proxy/lib/http-proxy/passes/web-incoming.js
++++ b/node_modules/http-proxy/lib/http-proxy/passes/web-incoming.js
+@@ -50,7 +50,7 @@ module.exports = {
+    */
+ 
+   timeout: function timeout(req, res, options) {
+-    if(options.timeout) {
++    if(options.timeout || options.timeout === 0) {
```
Comment (nevalla) [inline patches/http-proxy+1.17.0.patch]: 0 == false so we need to accept also exact 0 value

File: src/main/proxy.ts [hunk 2]
```diff
@@ -130,7 +130,8 @@ export class LensProxy {
     if (req.url.startsWith("/api-kube/")) {
       delete req.headers.authorization
       req.url = req.url.replace("/api-kube", "")
-      return await contextHandler.getApiTarget()
+      const timeout = req.url.includes("watch=") ? 0 : 30000
```
Comment (nevalla) [inline src/main/proxy.ts]: I hope this naive check is enough here.

File: src/main/context-handler.ts [hunk 3]
```diff
@@ -102,12 +102,15 @@ export class ContextHandler {
     }
   }
 
-  public async getApiTarget() {
-    if (this.apiTarget) { return this.apiTarget }
+  public async getApiTarget(timeout: number) {
+    if (this.apiTarget) {
+      this.apiTarget.timeout = timeout
```
Comment (jakolehm) [inline src/main/context-handler.ts]: I think this mutates `this.apiTarget` when first watch request comes in. Maybe safer to return always a fresh `apiTarget` object when request type is "watch"?

File: src/main/proxy.ts (hunk 2 above)
Comment (jakolehm) [reply src/main/proxy.ts]: We might be able to check request type from headers (watches are event streams).

File: patches/http-proxy+1.17.0.patch (hunk 1 above)
Comment (jakolehm) [reply patches/http-proxy+1.17.0.patch]: Could we do this without patching if we set big enough timeout value?

**Review by jakolehm — COMMENTED** [review]
Not completely sure what behaviour this fixes. Does this timeout happen only when watch request is "idling"?

File: patches/http-proxy+1.17.0.patch (hunk 1 above)
Comment (nevalla) [reply patches/http-proxy+1.17.0.patch]: Yes that's possible too. Any idea about what's big enough?

Comment (jakolehm) [reply patches/http-proxy+1.17.0.patch]: Few hours should be enough, see https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/ (–min-request-timeout).

**Comment by nevalla** [comment]
> Not completely sure what behaviour this fixes. Does this timeout happen only when watch request is "idling"?

Possibly yes. Docs says: 
> The number of milliseconds of inactivity before a socket is presumed to have timed out.

Found also this explanation:
> OK, but what does "inactivity on the socket" exactly mean? In a happy path, a TCP socket follows the following steps:
> 
> Start connecting
> DNS lookup is done: lookup event (Doesn't happen in HTTP Keep-Alive)
> Connection is made: connect event (Doesn't happen in HTTP Keep-Alive)
> Read data or write data
>
>When you call socket.setTimeout(), a timeout timer is created and restarted before connecting, after lookup, after connect and each data read & write.

https://shuheikagawa.com/blog/2017/05/13/http-request-timeouts-in-javascript/

I have a gut feeling that this is also happening on active watch requests, but I could be wrong too.

**Comment by nevalla** [comment]
I've dropped the `http-proxy` patch and using 4 hour timeout for watch requests now. Also for watch requests fresh `apiTarget` is always returned.

**Review by jakolehm — APPROVED** [review]

**Pull request merged**